# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

In [2]:
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)


children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.describe().T, data['children'].describe().T

(                    count           mean            std           min  \
 children          21525.0       0.538908       1.381587     -1.000000   
 days_employed     19351.0   63046.497661  140827.311974 -18388.949901   
 dob_years         21525.0      43.293380      12.574584      0.000000   
 education_id      21525.0       0.817236       0.548138      0.000000   
 family_status_id  21525.0       0.972544       1.420324      0.000000   
 debt              21525.0       0.080883       0.272661      0.000000   
 total_income      19351.0  167422.302208  102971.566448  20667.263793   
 
                             25%            50%            75%           max  
 children               0.000000       0.000000       1.000000  2.000000e+01  
 days_employed      -2747.423625   -1203.369529    -291.095954  4.017554e+05  
 dob_years             33.000000      42.000000      53.000000  7.500000e+01  
 education_id           1.000000       1.000000       1.000000  4.000000e+00  
 family_sta

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [1]:
data.isna().sum()

NameError: name 'data' is not defined

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [7]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

In [8]:
data['total_income'] = data['total_income'].fillna(data.groupby('income_type')['total_income'].median())
data['total_income'].isna().sum()

0

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [9]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [10]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [11]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [12]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [13]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [14]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [15]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [16]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [17]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [18]:
data.duplicated().sum()

71

In [19]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [20]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [21]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [22]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [23]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [24]:

data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [25]:
#находим общее количество должников, сгруппированных по кол-ву детей и делим  на всех заемщиков, сгруппированых по кол-ву детей 
debt_share_child = (data.groupby('children')['debt'].sum()) / (data.groupby('children')['children'].count()) 
#выводим долю должников (в процентах) по кол-ву детей. 
debt_share_child * 100
#Например: Из всех заемщиков с одним ребенком задолжности имеют приблизительно 9% 

    


children
0    7.543822
1    9.234609
2    9.454191
3    8.181818
4    9.756098
5    0.000000
dtype: float64

**Вывод:** Мы видим, что доля должников с разным количеством детей примерно равна, однако, у заемщиков с пятью детьми нет долгов. Каждый десятый заемщик имеющий 4х детей не платит кредит в срок. 

In [26]:
def categorize_child(row):
    try:
        if row > 0:
            return 'есть дети'    
        elif row == 0:
            return 'нет детей'
    except:
        return 'нет категории'

In [27]:
data['child_category'] = data['children'].apply(categorize_child)



In [28]:
#находим общее количество должников, сгруппированных по наличию детей и делим  на всех заемщиков, сгруппированых по наличию детей 
debt_share_child_category = (data.groupby('child_category')['debt'].sum()) / (data.groupby('child_category')['child_category'].count()) 
#выводим долю должников (в процентах) по наличию детей. 
debt_share_child_category * 100


child_category
есть дети    9.240331
нет детей    7.543822
dtype: float64

**Вывод:** Заемщики не имеющие детей с большей вероятностью выплатят кредит в срок!

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [29]:
#находим общее количество должников, сгруппированных по семейному положению и делим  на всех заемщиков, сгруппированых по семейному положению 
debt_share_fs = (data.groupby('family_status')['debt'].sum()) /(data.groupby('family_status')['family_status'].count())
#выводим долю должников (в процентах) по семейному положению, округляя до целых чисел. 
debt_share_fs * 100
#Например: Из всех заемщиков состоящих в браке задолжности имеют  приблизительно 8% 



family_status
Не женат / не замужем    9.763948
в разводе                7.064760
вдовец / вдова           6.624606
гражданский брак         9.313014
женат / замужем          7.560558
dtype: float64

**Вывод:** Наименьшую ответственность по займу имеют холостые заемщики. А более ответственно к займу относятся те заемщики, которые ранее состояли в браке - вдовцы и находящиеся в разводе. 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [30]:
#находим общее количество должников, сгруппированных по уровню дохода и делим  на всех заемщиков, сгруппированых по уровню дохода 
debt_share_income = (data.groupby('total_income_category')['debt'].sum()) / (data.groupby('total_income_category')['total_income_category'].count())
#выводим долю должников (в процентах) по уровню дохода, округляя до целых чисел. 
debt_share_income * 100
#Например: Из всех заемщиков c доходом до 30000  задолжности имеют приблизительно 9% 

total_income_category
A    8.000000
B    7.060231
C    8.498210
D    6.017192
E    9.090909
dtype: float64

**Вывод:** Самымим ответственными заемщиками являются те, чей доход находится в диапозоне  *30001 - 50000*. Большая доля должников выводится среди заемщиков с доходом *до 30000*. В остальных случаях доли примерно равны.

In [31]:
data.groupby('total_income_category').count()

children  days_employed  dob_years  education  \
total_income_category                                                  
A                            25             25         25         25   
B                          5014           5014       5014       5014   
C                         15921          15921      15921      15921   
D                           349            349        349        349   
E                            22             22         22         22   

                       education_id  family_status  family_status_id  gender  \
total_income_category                                                          
A                                25             25                25      25   
B                              5014           5014              5014    5014   
C                             15921          15921             15921   15921   
D                               349            349               349     349   
E                                22             22                22      22   

                       income_type   debt  total_income  purpose  \
total_income_category                                              
A                               25     25            25       25   
B                             5014   5014          5014     5014   
C                            15921  15921         15921    15921   
D                              349    349           349      349   
E                               22     22            22       22   

                       purpose_category  child_category  
total_income_category                                    
A                                    25              25  
B                                  5014            5014  
C                                 15921           15921  
D                                   349             349  
E                                    22              22

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [32]:
#находим общее количество должников, сгруппированных по целям и делим  на всех заемщиков, сгруппированых по целям  
debt_share_purpose = (data.groupby('purpose_category')['debt'].sum()) / (data.groupby('purpose_category')['purpose_category'].count())
#выводим долю должников (в процентах) по уровню дохода, округляя до целых чисел. 
debt_share_purpose * 100
#Например: Из всех заемщиков c целью проведения свадьбы задолжности имеют примерно 8% 


purpose_category
операции с автомобилем      9.347978
операции с недвижимостью    7.255139
получение образования       9.252758
проведение свадьбы          7.911803
dtype: float64

**Вывод:** Самые ответственные заемщики имеют дело с недвижимостью! Заемщики с целями "операции с автомобилем", "получение образования" реже возвращают кредит в срок.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
+ Пользователь пожелал не указывать некоторые сведения.
+ Автоматизированная система сбора информации дала сбой.
+ Невозможность получить данные.
+ Некорректное изменеие данных.
+ Ошибка разработчика

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медианное значение дает более объективную картину, так как не берет в расчет аномальные значения(слишком высокие или же слишком низкие, не соответствующие общей картине чисел). Если бы нам приводили не среднее значение зароботной платы в стране , а медианное, то все было бы наглядно и понятно, хотя и так понятно =)) 

### Шаг 4: общий вывод.

**Общий вывод:** 
 **Исходя из проведенного анализа можно привести следущие данные:**
- *Зависимость между количеством детей и возвратом кредита в срок (процент должников)*
    + 0---    7.543822
    + 1---   9.234609
    + 2---    9.454191
    + 3---    8.181818
    + 4---    9.756098
    + 5---    0.000000
- *Зависимость между наличием детей и возвратом кредита в срок (процент должников)*
    + есть дети---    9.240331
    + нет детей---    7.543822
- *Зависимость между семейным положением и возвратом кредита в срок (процент должников)*
    + Не женат / не замужем---    9.763948
    + в разводе---                7.064760
    + вдовец / вдова---           6.624606
    + гражданский брак---         9.313014
    + женат / замужем---          7.560558
- *Зависимость между уровнем дохода и возвратом кредита в срок (процент должников)*
    + 1000001 и выше---    8.000000
    + 200001–1000000---    7.060231
    + 50001–200000---     8.498210
    + 30001–50000---       6.017192
    + 0–30000---           9.090909 
- *Процент должников по целям кредита:* 
    + операции с автомобилем---      9.347978
    + операции с недвижимостью---    7.255139
    + получение образования---       9.252758
    + проведение свадьбы---          7.911803
   
- *С большей вероятностью проблемы с погашением кредита в срок будут у заемщиков, которые:*
    + зарабатывают меньше *30000*
    + берут займ с целью операции с автомобилем или получения образования 
    + не состоят в браке
    + имеют 4х детей



